forge/src/main/java/org/openjdk/skara/forge/bitbucket/BitbucketHost.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.skara.forge.bitbucket;
+
+import java.net.URI;
+import java.util.Optional;
+import org.openjdk.skara.forge.Forge;
+import org.openjdk.skara.forge.HostedCommit;
+import org.openjdk.skara.forge.HostedRepository;
+import org.openjdk.skara.host.Credential;
+import org.openjdk.skara.host.HostUser;
+import org.openjdk.skara.vcs.Hash;
+
+public class BitbucketHost implements Forge {
+    private final String name;
+    private final URI uri;
+    private final boolean useSsh;
+    private final int sshPort;
+    private final Credential credential;
+
+    public BitbucketHost(String name, URI uri, boolean useSsh, int sshPort, Credential credential) {
+        this.name = name;
+        this.uri = uri;
+        this.useSsh = useSsh;
+        this.sshPort = sshPort;
+        this.credential = credential;
+    }
+
+    public BitbucketHost(String name, URI uri, boolean useSsh, int sshPort) {
+        this.name = name;
+        this.uri = uri;
+        this.useSsh = useSsh;
+        this.sshPort = sshPort;
+        this.credential = null;
+    }
```

Comment:
It seems like credential cannot be `null` when `useSsh` is `true` (see the method `sshHostString`). Maybe the constructors should reflect this? Or use static factory methods like `BitBucketHost.withSSH` and/or `BitBucketHost.withPAT`?

Reply:
I more or less copy pasted the existing gitlab implementation here so this is following the same existing patterns. The same relationship between useSsh and credential are present there too. I agree it could be made neater with a builder pattern like most other factories do, or static factory methods.

Reply:
I decided to make it simpler by just removing the second constructor and let the factory deal with requiring credentials for ssh. Having two constructors didn't really make sense anymore as no API calls to bitbucket were needed.